wp/lib/bap_wp/src/output.ml

```diff
@@ -60,6 +60,45 @@ let format_model (model : Model.model) (env1 : Env.t) (_env2 : Env.t) : string =
   Format.flush_str_formatter ()
 
 
+type mem_model = {default : Constr.z3_expr ; model : (Constr.z3_expr * Constr.z3_expr) list}
+
+(** [extract_array] takes a z3 expression that is a seqeunce of store and converts it into
+    a mem_model, which consists of a key/value association list and a default value *)
+
+let extract_array (e : Constr.z3_expr) : mem_model  = 
+      let rec extract_array' (partial_map : (Constr.z3_expr * Constr.z3_expr) list) (e : Constr.z3_expr) : mem_model =
+          Printf.printf "%s\n" (Z3.Expr.to_string e);
+          let numargs = Z3.Expr.get_num_args e in
+          let args = Z3.Expr.get_args e in
+          let f_decl = Z3.Expr.get_func_decl e in
+          let f_name = Z3.FuncDecl.get_name f_decl |> Z3.Symbol.to_string in
+          if ( Int.(numargs = 3) && String.(f_name = "store"))
+          then begin 
+                let next_arr = List.nth_exn args 0 in
+                let key = List.nth_exn args 1 in
+                let value = List.nth_exn args 2 in
+                extract_array' (( key , value ) :: partial_map) next_arr
+                end
+          else if ( Int.(numargs = 1) && String.(f_name = "const")) then begin
+                let key = List.nth_exn args 0 in
+                { default = key; model = List.rev partial_map}
+                end
+          else begin
+                warning "Unexpected case destructing Z3 array: %s" f_name;
+                {default = e ; model = partial_map}
+                end
+      in
+      extract_array' [] e
+
+let get_mem (m : Z3.Model.model) : mem_model option =
+   let decls = Z3.Model.get_decls m |> List.filter ~f:(fun decl -> (Z3.Symbol.to_string (Z3.FuncDecl.get_name decl)) = "mem0") in
+   match (List.hd decls) with
+   | None -> warning "No mem0 found.";
```

Comment:
I wonder if there's a way to be more agnostic about the name of the `mem` variable, e.g. by traipsing around a `CPU` or `Target` instance to be able to call `CPU.mem` from here: http://binaryanalysisplatform.github.io/bap/api/v1.6.0/Bap.Std.CPU.html

And get the name for the memory variable.

Reply:
See e.g. what we do here: https://github.com/draperlaboratory/cbat_tools/blob/9f1180be9e01c4d51c5ab203494c062d6ef1c291/wp/lib/bap_wp/src/precondition.ml#L260